scalding-serialization-macros/src/main/scala/com/twitter/scalding/serialization/macros/impl/ordered_serialization/SealedTraitLike.scala

```diff
@@ -0,0 +1,323 @@
+/*
+ Copyright 2014 Twitter, Inc.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+package com.twitter.scalding.serialization.macros.impl.ordered_serialization
+
+import scala.language.experimental.macros
+import scala.reflect.macros.Context
+
+import com.twitter.scalding._
+import com.twitter.scalding.serialization.OrderedSerialization
+
+object SealedTraitLike {
+
+  def compareBinary(c: Context)(inputStreamA: c.TermName, inputStreamB: c.TermName)(subData: List[(Int, c.Type, Option[TreeOrderedBuf[c.type]])]): c.Tree = {
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+    val valueA = freshT("valueA")
+    val valueB = freshT("valueB")
+    val idxCmp = freshT("idxCmp")
+
+    val compareSameTypes: Tree = subData.foldLeft(Option.empty[Tree]) {
+      case (existing, (idx, tpe, optiTBuf)) =>
+
+        val commonCmp: Tree = optiTBuf.map{ tBuf =>
+          tBuf.compareBinary(inputStreamA, inputStreamB)
+        }.getOrElse[Tree](q"0")
+
+        existing match {
+          case Some(t) =>
+            Some(q"""
+              if($valueA == $idx) {
+                $commonCmp
+              } else {
+                $t
+              }
+            """)
+          case None =>
+            Some(q"""
+                if($valueA == $idx) {
+                  $commonCmp
+                } else {
+                  sys.error("Unable to compare unknown type")
```

Comment:
I guess the way you are building this, it will be the most nested thing and it can only be reached if somehow none of the ids match, which should be impossible.

If that is right can you put a different message, something like: `unreachable code`.

Also, I wonder if it makes sense to write code to do a binary search rather than a linear one as is done here.